fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -3375,57 +3383,81 @@ private boolean areAllRecordTypesSince(@Nullable Collection<RecordType> recordTy
         final boolean newStore = oldFormatVersion == 0;
         final Map<Index, List<RecordType>> indexes = metaData.getIndexesSince(oldMetaDataVersion);
         if (!indexes.isEmpty()) {
-            // Can only determine overall emptiness (without an additional read) if got count for all record types.
-            Pair<CompletableFuture<Long>, Boolean> counterAndIsAllTypes = getRecordCountForRebuildIndexes(newStore, rebuildRecordCounts, indexes);
-            return counterAndIsAllTypes.getLeft().thenCompose(recordCount -> {
-                if (recordCount == 0 && counterAndIsAllTypes.getRight() &&
-                        formatVersion >= SAVE_UNSPLIT_WITH_SUFFIX_FORMAT_VERSION && omitUnsplitRecordSuffix) {
-                    // There are no records. Don't use the legacy split format.
-                    if (newStore ? LOGGER.isDebugEnabled() : LOGGER.isInfoEnabled()) {
-                        KeyValueLogMessage msg = KeyValueLogMessage.build("upgrading unsplit format on empty store",
-                                                    LogMessageKeys.NEW_FORMAT_VERSION, formatVersion,
-                                                    subspaceProvider.logKey(), subspaceProvider.toString(context));
-                        if (newStore) {
-                            LOGGER.debug(msg.toString());
-                        } else {
-                            LOGGER.info(msg.toString());
+            // If all the new indexes are only for a record type whose primary key has a type prefix, then we can scan less.
+            RecordType singleRecordTypeWithPrefixKey = singleRecordTypeWithPrefixKey(indexes);
```

Comment:
This isn't about the cardinality of any of the types, but whether the primary keys are contiguous. It might be possible for multiple record types to be, if you assume that the type keys are dense.